README.md
---
title: Device Orientation
description: Access compass data.
---
<!---
# license: Licensed to the Apache Software Foundation (ASF) under one
#         or more contributor license agreements.  See the NOTICE file
#         distributed with this work for additional information
#         regarding copyright ownership.  The ASF licenses this file
#         to you under the Apache License, Version 2.0 (the
#         "License"); you may not use this file except in compliance
#         with the License.  You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#         Unless required by applicable law or agreed to in writing,
#         software distributed under the License is distributed on an
#         "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#         KIND, either express or implied.  See the License for the
#         specific language governing permissions and limitations
#         under the License.
-->

# cordova-plugin-device-orientation

[![Android Testsuite](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/android.yml/badge.svg)](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/android.yml) [![Chrome Testsuite](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/chrome.yml/badge.svg)](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/chrome.yml) [![iOS Testsuite](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/ios.yml/badge.svg)](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/ios.yml) [![Lint Test](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/lint.yml/badge.svg)](https://github.com/apache/cordova-plugin-device-orientation/actions/workflows/lint.yml)

# Usage Notice

With the [W3C Device Orientation API](https://www.w3.org/TR/orientation-event/), Android, iOS, and Windows devices may not need this plugin anymore.

However, on iOS 13+, potential issues with permissions and secure contexts can arise. Therefore it is recommended to use this plugin as it uses a native implementation.

------

## Description

This plugin provides access to the device's compass. The compass is a sensor
that detects the direction or heading that the device is pointed, typically
from the top of the device.  It measures the heading in degrees from 0 to
359.99, where 0 is north.

Access is via a global `navigator.compass` object.

Although the object is attached to the global scoped `navigator`, it is not available until after the `deviceready` event.

    document.addEventListener("deviceready", onDeviceReady, false);
    function onDeviceReady() {
        console.log(navigator.compass);
    }

Report issues on the [Apache Cordova issue tracker](https://issues.apache.org/jira/secure/RapidBoard.jspa?rapidView=190&projectKey=CB)

## Installation

    cordova plugin add cordova-plugin-device-orientation

## Supported Platforms

- Android
- Browser
- iOS
- Windows

## Methods

- navigator.compass.getCurrentHeading
- navigator.compass.watchHeading
- navigator.compass.clearWatch

## navigator.compass.getCurrentHeading

Get the current compass heading. The compass heading is returned via a `CompassHeading`
object using the `compassSuccess` callback function.

    navigator.compass.getCurrentHeading(compassSuccess, compassError);

### Example

    function onSuccess(heading) {
        alert('Heading: ' + heading.magneticHeading);
    };

    function onError(error) {
        alert('CompassError: ' + error.code);
    };

    navigator.compass.getCurrentHeading(onSuccess, onError);

## navigator.compass.watchHeading

Gets the device's current heading at a regular interval. Each time the heading
is retrieved, the `headingSuccess` callback function is executed.

The returned watch ID references the compass watch interval. The watch
ID can be used with `navigator.compass.clearWatch` to stop watching the navigator.compass.

    var watchID = navigator.compass.watchHeading(compassSuccess, compassError, [compassOptions]);

`compassOptions` may contain the following keys:

- __frequency__: How often to retrieve the compass heading in milliseconds. _(Number)_ (Default: 100)
- __filter__: The change in degrees required to initiate a watchHeading success callback. When this value is set, __frequency__ is ignored. _(Number)_

### Example

    function onSuccess(heading) {
        var element = document.getElementById('heading');
        element.innerHTML = 'Heading: ' + heading.magneticHeading;
    };

    function onError(compassError) {
        alert('Compass error: ' + compassError.code);
    };

    var options = {
        frequency: 3000
    }; // Update every 3 seconds

    var watchID = navigator.compass.watchHeading(onSuccess, onError, options);

### Update 29/08/2022

On Android, the plugin used a `Sensor.TYPE_ORIENTATION` which is [deprecated](https://developer.android.com/guide/topics/sensors/sensors_position).
This sensor was more and more [omitted](https://github.com/apache/cordova-plugin-device-orientation/issues/64) in new devices.
So now the plugin uses `Sensor.TYPE_ACCELEROMETER` and `Sensor.TYPE_MAGNETIC_FIELD` which are available everywhere.
They are fused together as instructed [here](https://developer.android.com/guide/topics/sensors/sensors_position#sensors-pos-orient),
except that the SensorManager.getOrientation() has a [bug that will not be fixed](https://issuetracker.google.com/issues/37127944).
[Stochastically](https://stackoverflow.com/users/2110762/stochastically) found a [solution](https://stackoverflow.com/questions/15537125/inconsistent-orientation-sensor-values-on-android-for-azimuth-yaw-and-roll/16418016#16418016) that works like a charm and is used here.


### Browser Quirks

Values for current heading are randomly generated in order to simulate the compass.

### iOS Quirks

Only one `watchHeading` can be in effect at one time in iOS.  If a
`watchHeading` uses a filter, calling `getCurrentHeading` or
`watchHeading` uses the existing filter value to specify heading
changes. Watching heading changes with a filter is more efficient than
with time intervals.

### Android Quirks

- No support for `filter`.

## navigator.compass.clearWatch

Stop watching the compass referenced by the watch ID parameter.

    navigator.compass.clearWatch(watchID);

- __watchID__: The ID returned by `navigator.compass.watchHeading`.

### Example

    var watchID = navigator.compass.watchHeading(onSuccess, onError, options);

    // ... later on ...

    navigator.compass.clearWatch(watchID);

## CompassHeading

A `CompassHeading` object is returned to the `compassSuccess` callback function.

### Properties

- __magneticHeading__: The heading in degrees from 0-359.99 at a single moment in time. _(Number)_

- __trueHeading__: The heading relative to the geographic North Pole in degrees 0-359.99 at a single moment in time. A negative value indicates that the true heading can't be determined.  _(Number)_

- __headingAccuracy__: The deviation in degrees between the reported heading and the true heading. _(Number)_

- __timestamp__: The time at which this heading was determined.  _(DOMTimeStamp)_

### Android Quirks

- The `trueHeading` property is not supported, but reports the same value as `magneticHeading`.

- The `headingAccuracy` property is always 0 because there is no difference between the `magneticHeading` and `trueHeading`.

### iOS Quirks

- The `trueHeading` property is only returned for location services enabled via `navigator.geolocation.watchLocation()`.

## CompassError

A `CompassError` object is returned to the `compassError` callback function when an error occurs.

### Properties

- __code__: One of the predefined error codes listed below.

### Constants

- `CompassError.COMPASS_INTERNAL_ERR`
- `CompassError.COMPASS_NOT_SUPPORTED`


RELEASENOTES.md
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
# http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->
# Release Notes

### 3.0.0 (Oct 27, 2023)

Breaking:

* [GH-57](https://github.com/apache/cordova-plugin-device-orientation/pull/57) fix!: remove deprecated platforms

**Chores:**

* [GH-83](https://github.com/apache/cordova-plugin-device-orientation/pull/83) chore: add asf config
* [GH-76](https://github.com/apache/cordova-plugin-device-orientation/pull/76) chore(npm): bump package-lock v2 w/ rebuild
* [GH-62](https://github.com/apache/cordova-plugin-device-orientation/pull/62) chore: adds `package-lock` file
* [GH-58](https://github.com/apache/cordova-plugin-device-orientation/pull/58) chore: package cleanup
* [GH-55](https://github.com/apache/cordova-plugin-device-orientation/pull/55) chore(npm): adds ignore list

**Others:**

* [GH-84](https://github.com/apache/cordova-plugin-device-orientation/pull/84) dep: bump `word-wrap` from `1.2.3` to `1.2.5`
* [GH-79](https://github.com/apache/cordova-plugin-device-orientation/pull/79) dep: bump `json5` from `1.0.1` to `1.0.2`
* [GH-77](https://github.com/apache/cordova-plugin-device-orientation/pull/77) ci: sync workflow with paramedic
* [GH-70](https://github.com/apache/cordova-plugin-device-orientation/pull/70) ci(android): update `java` requirement for `cordova-android`@11
* [GH-67](https://github.com/apache/cordova-plugin-device-orientation/pull/67) ci(ios): update workflow w/ **iOS** 15
* [GH-66](https://github.com/apache/cordova-plugin-device-orientation/pull/66) ci: remove old ci service
* [GH-65](https://github.com/apache/cordova-plugin-device-orientation/pull/65) ci: add github actions workflow
* [GH-61](https://github.com/apache/cordova-plugin-device-orientation/pull/61) refactor(eslint): use cordova-eslint w/ fix
* [GH-56](https://github.com/apache/cordova-plugin-device-orientation/pull/56) docs: remove outdated translations
* [GH-54](https://github.com/apache/cordova-plugin-device-orientation/pull/54) docs: undeprecate plugin

### 2.0.1 (April 13, 2017)
* [CB-14002](https://issues.apache.org/jira/browse/CB-14002): Fix `cordovaDependencies` to allow plugin install 

### 2.0.0 (Sep 18, 2017)
* [CB-13076](https://issues.apache.org/jira/browse/CB-13076) added deprecation notice to info tag
* [CB-12728](https://issues.apache.org/jira/browse/CB-12728) Device Orientation - SUNSET
* [CB-13028](https://issues.apache.org/jira/browse/CB-13028) (CI) **Browser** builds on `Travis` and `AppVeyor`
* [CB-12667](https://issues.apache.org/jira/browse/CB-12667) Reset changes for searching Samsung sensors
* [CB-13000](https://issues.apache.org/jira/browse/CB-13000) (CI) Speed up **Android** builds
* [CB-12991](https://issues.apache.org/jira/browse/CB-12991) (CI) Updated CI badges
* [CB-12935](https://issues.apache.org/jira/browse/CB-12935) (**windows**) Enable paramedic builds on `AppVeyor`
* [CB-12935](https://issues.apache.org/jira/browse/CB-12935) (**ios**, **Android**) Enable paramedic builds on `Travis CI`
* [CB-12667](https://issues.apache.org/jira/browse/CB-12667) **Android**: Added logic for searching sensors from Samsung vendor
* [CB-12847](https://issues.apache.org/jira/browse/CB-12847) added `bugs` entry to `package.json`.

### 1.0.7 (Apr 27, 2017)
* [CB-12622](https://issues.apache.org/jira/browse/CB-12622) Added **Android 6.0** build badge to `README`
* [CB-12685](https://issues.apache.org/jira/browse/CB-12685) added `package.json` to tests folder

### 1.0.6 (Feb 28, 2017)
* [CB-12353](https://issues.apache.org/jira/browse/CB-12353) Corrected merges usage in `plugin.xml`
* [CB-12369](https://issues.apache.org/jira/browse/CB-12369) Add plugin typings from `DefinitelyTyped`
* [CB-12363](https://issues.apache.org/jira/browse/CB-12363) Added build badges for **iOS 9.3** and **iOS 10.0**
* [CB-12230](https://issues.apache.org/jira/browse/CB-12230) Removed **Windows 8.1** build badges

### 1.0.5 (Dec 07, 2016)
* [CB-12224](https://issues.apache.org/jira/browse/CB-12224) Updated version and RELEASENOTES.md for release 1.0.5
* [CB-9179](https://issues.apache.org/jira/browse/CB-9179) (ios) Fixed trueHeading being always 0
* [CB-11917](https://issues.apache.org/jira/browse/CB-11917) - Remove pull request template checklist item: "iCLA has been submitted…"
* [CB-11832](https://issues.apache.org/jira/browse/CB-11832) Incremented plugin version.

### 1.0.4 (Sep 08, 2016)
* [CB-11795](https://issues.apache.org/jira/browse/CB-11795) Add 'protective' entry to cordovaDependencies
* Add badges for paramedic builds on Jenkins
* Add pull request template.
* [CB-10996](https://issues.apache.org/jira/browse/CB-10996) Adding front matter to README.md
* Removed ios quirk code + documentation

### 1.0.3 (Apr 15, 2016)
* Remove `warning` emoji, as it doesn't correctly display in the docs website: http://cordova.apache.org/docs/en/dev/cordova-plugin-device-orientation/index.html
* [CB-10636](https://issues.apache.org/jira/browse/CB-10636) Add `JSHint` for plugins

### 1.0.2 (Nov 18, 2015)
* [CB-10035](https://issues.apache.org/jira/browse/CB-10035) Updated `RELEASENOTES` to be newest to oldest
* [CB-4596](https://issues.apache.org/jira/browse/CB-4596) Fix `timestamp` to be `DOMTimeStamp` across the board
* Fixing contribute link.

src/android/CompassListener.java
/*
       Licensed to the Apache Software Foundation (ASF) under one
       or more contributor license agreements.  See the NOTICE file
       distributed with this work for additional information
       regarding copyright ownership.  The ASF licenses this file
       to you under the Apache License, Version 2.0 (the
       "License"); you may not use this file except in compliance
       with the License.  You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

       Unless required by applicable law or agreed to in writing,
       software distributed under the License is distributed on an
       "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
       KIND, either express or implied.  See the License for the
       specific language governing permissions and limitations
       under the License.
*/
package org.apache.cordova.deviceorientation;

import static java.lang.Float.isNaN;

import java.util.List;

import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.content.Context;

import android.os.Handler;
import android.os.Looper;

import org.apache.cordova.LOG;

/**
 * This class listens to the compass sensor and stores the latest heading value.
 */
public class CompassListener extends CordovaPlugin implements SensorEventListener {

  public static int STOPPED = 0;
  public static int STARTING = 1;
  public static int RUNNING = 2;
  public static int ERROR_FAILED_TO_START = 3;

  public long TIMEOUT = 30000; // Timeout in msec to shut off listener

  int status; // status of listener
  // float heading; // most recent heading value
  // Replaced by sin & cos
  float sinHeading; // sin of most recent heading value
  float cosHeading; // cos of most recent heading value
  long timeStamp; // time of most recent value
  long lastAccessTime; // time the value was last retrieved
  int rotationAccuracy = 2; // SENSOR_STATUS_ACCURACY_MEDIUM accuracy of the sensor

  private SensorManager sensorManager;// Sensor manager

  Sensor rvSensor; // Compass sensor returned by sensor manager

  private CallbackContext callbackContext;

  /**
   * Constructor.
   */
  public CompassListener() {
    this.sinHeading = 0;
    this.cosHeading = 0;
    this.timeStamp = 0;
    this.setStatus(CompassListener.STOPPED);
  }

  /**
   * Sets the context of the Command. This can then be used to do things like
   * get file paths associated with the Activity.
   *
   * @param cordova The context of the main Activity.
   * @param webView The CordovaWebView Cordova is running in.
   */
  public void initialize(CordovaInterface cordova, CordovaWebView webView) {
    super.initialize(cordova, webView);
    this.sensorManager = (SensorManager) cordova.getActivity().getSystemService(Context.SENSOR_SERVICE);
  }

  /**
   * Executes the request and returns PluginResult.
   *
   * @param action            The action to execute.
   * @param args              JSONArry of arguments for the plugin.
   * @param callbackS=Context The callback id used when calling back into
   *                          JavaScript.
   * @return True if the action was valid.
   * @throws JSONException
   */
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    if (action.equals("start")) {
      this.start();
    } else if (action.equals("stop")) {
      this.stop();
    } else if (action.equals("getStatus")) {
      int i = this.getStatus();
      callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, i));
    } else if (action.equals("getHeading")) {
      // If not running, then this is an async call, so don't worry about waiting
      if (this.status != CompassListener.RUNNING) {
        int r = this.start();
        if (r == CompassListener.ERROR_FAILED_TO_START) {
          callbackContext.sendPluginResult(
              new PluginResult(PluginResult.Status.IO_EXCEPTION, CompassListener.ERROR_FAILED_TO_START));
          return true;
        }
        // Set a timeout callback on the main thread.
        Handler handler = new Handler(Looper.getMainLooper());
        handler.postDelayed(new Runnable() {
          public void run() {
            CompassListener.this.timeout();
          }
        }, 2000);
      }
      callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, getCompassHeading()));
    } else if (action.equals("setTimeout")) {
      this.setTimeout(args.getLong(0));
    } else if (action.equals("getTimeout")) {
      long l = this.getTimeout();
      callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, l));
    } else {
      // Unsupported action
      return false;
    }
    return true;
  }

  /**
   * Called when listener is to be shut down and object is being destroyed.
   */
  public void onDestroy() {
    this.stop();
  }

  /**
   * Called when app has navigated and JS listeners have been destroyed.
   */
  public void onReset() {
    this.stop();
  }

  // --------------------------------------------------------------------------
  // LOCAL METHODS
  // --------------------------------------------------------------------------

  /**
   * Start listening for compass sensor.
   *
   * @return status of listener
   */
  public int start() {
    // final int SENSOR_DELAY = sensorManager.SENSOR_DELAY_NORMAL;
    final int SENSOR_DELAY = sensorManager.SENSOR_DELAY_UI;
    // final int SENSOR_DELAY = sensorManager.SENSOR_DELAY_GAME;

    // If already starting or running, then just return
    if ((this.status == CompassListener.RUNNING) || (this.status == CompassListener.STARTING)) {
      return this.status;
    }

    // use accelerometer & magnetometer
    // http://web.archive.org/web/20151205103652/http://www.codingforandroid.com/2011/01/using-orientation-sensors-simple.html
    // https://android-developers.googleblog.com/2010/09/one-screen-turn-deserves-another.html
    // https://stackoverflow.com/questions/15537125/inconsistent-orientation-sensor-values-on-android-for-azimuth-yaw-and-roll/16418016#16418016
    rvSensor = this.sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
    if (rvSensor != null) {
      this.sensorManager.registerListener(this, this.rvSensor, SENSOR_DELAY);
      this.lastAccessTime = System.currentTimeMillis();
      this.setStatus(CompassListener.STARTING);

    } else {
      // If error, then set status to error
      this.setStatus(CompassListener.ERROR_FAILED_TO_START);
    }
    return this.status;
  }

  /**
   * Stop listening to compass sensor.
   */
  public void stop() {
    if (this.status != CompassListener.STOPPED) {
      this.sensorManager.unregisterListener(this);
    }
    this.setStatus(CompassListener.STOPPED);
  }

  public void onAccuracyChanged(Sensor sensor, int accuracy) {
    // TODO Auto-generated method stub
  }

  /**
   * Called after a delay to time out if the listener has not attached fast
   * enough.
   */
  private void timeout() {
    if (this.status == CompassListener.STARTING) {
      this.setStatus(CompassListener.ERROR_FAILED_TO_START);
      if (this.callbackContext != null) {
        this.callbackContext.error("Compass listener failed to start.");
      }
    }
  }

  private float rotationVector[] = null;

  private float toDeg(float x) {
    return x * 180 / (float) Math.PI; // Convert to degrees
  }

  /**
   * Sensor listener event.
   *
   * @param SensorEvent event
   */
  @SuppressWarnings("deprecation")
  public void onSensorChanged(SensorEvent event) {
    float sinHeading = (0f / 0f); // NaN
    float cosHeading = (0f / 0f); // NaN

    long myNow = System.currentTimeMillis();
    switch (event.sensor.getType()) {
      case Sensor.TYPE_ROTATION_VECTOR:
        if (rotationVector == null) {
          rotationVector = new float[3];
        }
        rotationVector = event.values;
        this.rotationAccuracy = event.accuracy;
        break;
    }

    if (rotationVector != null) {
      float R[] = new float[9];
      SensorManager.getRotationMatrixFromVector(R, event.values);
      // Android recommends using SensorManager.getOrientation() but it has a wontFix
      // bug:
      // https://stackoverflow.com/questions/67824884/pitch-returned-by-getorientation-function-is-wrong
      // So we use Stochastically's method:
      // https://stackoverflow.com/questions/15537125/inconsistent-orientation-sensor-values-on-android-for-azimuth-yaw-and-roll/16418016#16418016
      // which works like a charm.
      // Beware also of screen orientation:
      // https://android-developers.googleblog.com/2010/09/one-screen-turn-deserves-another.html

      // heading = (float) Math.atan2((double) (R[1] - R[3]), (double) (R[0] + R[4]));
      // heading = toDeg(heading);
      // Replaced by sin & cos
      this.sinHeading = R[1] - R[3];
      this.cosHeading = R[0] + R[4];
    }
    this.timeStamp = System.currentTimeMillis();
    this.setStatus(CompassListener.RUNNING);

    // If heading hasn't been read for TIMEOUT time, then turn off compass sensor to
    // save power
    if ((this.timeStamp - this.lastAccessTime) > this.TIMEOUT) {
      this.stop();
    }
  }

  /**
   * Get status of compass sensor.
   *
   * @return status
   */
  public int getStatus() {
    return this.status;
  }

  /**
   * Set the timeout to turn off compass sensor if getHeading() hasn't been
   * called.
   *
   * @param timeout Timeout in msec.
   */
  public void setTimeout(long timeout) {
    this.TIMEOUT = timeout;
  }

  /**
   * Get the timeout to turn off compass sensor if getHeading() hasn't been
   * called.
   *
   * @return timeout in msec
   */
  public long getTimeout() {
    return this.TIMEOUT;
  }

  /**
   * Set the status and send it to JavaScript.
   *
   * @param status
   */
  private void setStatus(int status) {
    this.status = status;
  }

  /**
   * Create the CompassHeading JSON object to be returned to JavaScript
   *
   * @return a compass heading
   */
  private JSONObject getCompassHeading() throws JSONException {
    JSONObject obj = new JSONObject();
    long myNow = System.currentTimeMillis();

    // obj.put("magneticHeading", this.heading);
    // obj.put("trueHeading", this.heading);
    obj.put("sinHeading", this.sinHeading);
    obj.put("cosHeading", this.cosHeading);
    // Since the magnetic and true heading are always the same our and accuracy
    // is defined as the difference between true and magnetic always return zero
    obj.put("headingAccuracy", rotationAccuracy);
    obj.put("timeStamp", myNow);
    this.lastAccessTime = myNow;
    return obj;
  }
}
